Automattic/newspack-newsletters
# feat: pixel and click-tracking

### All Submissions:






### Changes proposed in this Pull Request:

Implements pixel and click-tracking for sent newsletters.

The pixel is appended to the body of every sent email and increments a post meta as well as sends out a [data event](https://github.com/Automattic/newspack-plugin/blob/08e53ff639ffe3c58573b4f368f1db7285ff4055/includes/data-events/README.md) called `newsletter_opened` with the newsletter ID and email address.

The click tracker filters the existing `newspack_newsletters_process_link` to replace the URL with a proxied version. The redirection handler increments a post meta and sends a data event named `newsletter_clicked` with the newsletter ID, email address, and destination URL.

Both data events should be listed in the Connections -> Webhooks while editing/adding a webhook. (with `define( 'NEWSPACK_EXPERIMENTAL_WEBHOOKS', true );`)

Opens and clicks are also added as sortable columns in the dashboard's newsletters table:

<img width="1540" alt="image" src="https://github.com/Automattic/newspack-newsletters/assets/820752/3dfe56ec-61ec-4ee2-ae8a-cff75d434816">

**NOTE: This branch will be merged to `epic/revenue-optimizations` to be released along with other enhancements.**

### How to test the changes in this Pull Request:

Written tests should pass

1. Check out this branch and send a newsletter with a link to `https://newspack.com/learn-more/?v=2#post-3415` and another to any other destination
2. Open the newsletter in your email client (make sure to allow images to load), visit the dashboard, and confirm "Opened" column shows `1` for this newsletter
3. Click on a link from the newsletter and confirm you are redirected to the correct destination (with hash and parameters)
4. Refresh the dashboard and confirm the "Clicks" count is 1
5. Click on the other link and confirm it's now 2
6. Click again on the same link and confirm it's 3 (clicks are not uniquely identified)
7. Send another newsletter, open and click on links as well
8. Refresh the dashboard and confirm you are able to sort by opens and clicks

### Other information:
